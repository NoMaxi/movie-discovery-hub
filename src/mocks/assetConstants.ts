// Mock version of `src/constants/assetConstants.ts` for tests (mock images imports)
export const DEFAULT_IMAGE_URL = "test-image-stub";
