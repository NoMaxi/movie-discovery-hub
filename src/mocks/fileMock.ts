// Stub for image file imports in Jest tests
export default "";
