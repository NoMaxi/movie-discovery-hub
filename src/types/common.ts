export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
