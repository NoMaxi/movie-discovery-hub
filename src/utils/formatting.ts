export const formatDuration = (totalMinutes: number): string => {
    if (!Number.isFinite(totalMinutes) || totalMinutes < 0) {
        return "";
    }

    if (totalMinutes === 0) {
        return "0min";
    }

    const hours = Math.floor(totalMinutes / 60);
    const minutes = Math.round(totalMinutes % 60);

    const parts = [hours > 0 && `${hours}h`, minutes > 0 && `${minutes}min`].filter(Boolean);

    return parts.join(" ");
};
