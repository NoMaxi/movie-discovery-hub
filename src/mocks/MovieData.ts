import { InitialMovieInfo, MovieDetailsData } from "@/types/common";

export const mockMovieDetails: MovieDetailsData = {
    id: "1",
    title: "Pulp Fiction",
    releaseYear: 1994,
    genres: ["Comedy", "Crime"],
    imageUrl: "https://upload.wikimedia.org/wikipedia/en/3/3b/Pulp_Fiction_%281994%29_poster.jpg",
    rating: 8.9,
    duration: 154,
    description:
        "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
};

export const alternativeMovieDetails: MovieDetailsData = {
    id: "2",
    imageUrl: "https://example.com/poster-alt.jpg",
    title: "Another Great Film",
    releaseYear: 2022,
    genres: ["Comedy", "Crime"],
    rating: 7.9,
    duration: 59.5,
    description: "The greatest movie in history of cinema.",
};

export const mockMovieInfo: InitialMovieInfo = {
    ...mockMovieDetails,
    releaseDate: "1994-10-14",
};
