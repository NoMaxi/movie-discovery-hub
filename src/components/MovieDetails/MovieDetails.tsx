import { Movie } from "@/types/common";
import { formatDuration } from "@/utils/formatting";
import noPosterImage from "@/assets/no-poster-image.png";
import { useImageFallback } from "@/hooks/useImageFallback/useImageFallback";

interface MovieDetailsProps {
    details: Movie;
}

export const MovieDetails = ({
    details: { description, duration, genres, imageUrl, rating, releaseYear, title },
}: MovieDetailsProps) => {
    const imageProps = useImageFallback(imageUrl, noPosterImage);
    return (
        <div
            data-testid="movie-details"
            className="
                movie-details flex w-full max-h-header-height-expanded gap-x-14 p-8 text-text
                bg-content-background
            "
        >
            <img
                src={imageProps.src}
                alt={`${title} poster`}
                className="movie-details-poster flex-shrink-0 w-[320px] h-[485px] object-cover"
                onError={imageProps.onError}
            />

            <div className="movie-details-info flex flex-col flex-grow">
                <div className="flex items-center gap-x-5 mb-2">
                    <h1 className="movie-details-title text-2xl font-light tracking-[1px] uppercase">{title}</h1>{" "}
                    <div
                        className="
                            movie-details-rating flex flex-shrink-0 items-center justify-center
                            w-3xl h-3xl border-2 rounded-full
                        "
                    >
                        <span className="font-light">{rating}</span>
                    </div>
                </div>

                <p className="movie-details-genres text-md opacity-50 mb-6">{genres.join(", ")}</p>

                <div className="flex items-center gap-x-6 text-[24px] font-light text-primary mb-6">
                    <span className="movie-details-date">{releaseYear}</span>
                    <span className="movie-details-duration">{formatDuration(duration)}</span>
                </div>

                <p
                    className="
                    movie-details-description font-light opacity-50 leading-relaxed overflow-y-scroll"
                >
                    {description}
                </p>
            </div>
        </div>
    );
};
