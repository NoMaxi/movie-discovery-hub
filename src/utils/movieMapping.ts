import { InitialMovieInfo, Movie, SortOption } from "@/types/common";
import { DEFAULT_DESCRIPTION, DEFAULT_TITLE } from "@/constants/constants";
import { DEFAULT_IMAGE_URL } from "@/constants/assetConstants";
import { getYearFromDate } from "@/utils/formatting";
import { APIMovieDetails, SortBy, SortOrder } from "@/services/movieService";

export const mapAPIMovieDetailsToMovie = ({
    id,
    genres,
    overview,
    poster_path,
    release_date,
    runtime,
    title,
    vote_average,
}: APIMovieDetails): Movie => ({
    id,
    description: overview || DEFAULT_DESCRIPTION,
    duration: runtime ?? 0,
    genres,
    imageUrl: poster_path || DEFAULT_IMAGE_URL,
    rating: vote_average ?? 0,
    releaseYear: getYearFromDate(release_date),
    title: title || DEFAULT_TITLE,
});

export const mapAPIMovieDetailsToInitialMovieInfo = (apiMovieDetails: APIMovieDetails): InitialMovieInfo => {
    const movie = mapAPIMovieDetailsToMovie(apiMovieDetails);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { releaseYear, ...movieWithoutReleaseYear } = movie;

    return {
        ...movieWithoutReleaseYear,
        releaseDate: apiMovieDetails.release_date ?? "",
    };
};

const SORT_OPTION_TO_SORT_BY: Record<SortOption, SortBy> = {
    "Release Date": "release_date",
    Title: "title",
};

const SORT_OPTION_TO_SORT_ORDER: Record<SortOption, SortOrder> = {
    "Release Date": "desc",
    Title: "asc",
};

export const mapSortOptionToSortBy = (sortOption: SortOption): SortBy => {
    return SORT_OPTION_TO_SORT_BY[sortOption] || "release_date";
};

export const mapSortOptionToSortOrder = (sortOption: SortOption): SortOrder => {
    return SORT_OPTION_TO_SORT_ORDER[sortOption] || "desc";
};
