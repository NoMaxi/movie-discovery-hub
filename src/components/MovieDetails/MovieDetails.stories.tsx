import type { Meta, StoryObj } from "@storybook/react";
import MovieDetails from "./MovieDetails";
import { mockMovieDetails } from "./MovieDetails.mocks";

const meta = {
    component: MovieDetails,
    argTypes: {
        details: {
            description: "Movie details data object",
            control: { type: "object" },
        },
    },
} satisfies Meta<typeof MovieDetails>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        details: mockMovieDetails,
    },
};

export const ZeroDurationMovie: Story = {
    args: {
        details: { ...mockMovieDetails, duration: 0 },
    },
};

