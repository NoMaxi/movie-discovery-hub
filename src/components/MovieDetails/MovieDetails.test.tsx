import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { formatDuration } from "@/utils/formatting";
import { MovieDetailsData } from "@/types/common";
import { mockMovieDetails, alternativeMovieDetails } from "@/mocks/MovieData";
import { MovieDetails } from "./MovieDetails";

describe("MovieDetails", () => {
    const renderComponent = (details: MovieDetailsData) => {
        return render(<MovieDetails details={details} />);
    };

    test("Should render movie details correctly with provided data", () => {
        const { asFragment } = renderComponent(mockMovieDetails);
        const img = screen.getByAltText(`${mockMovieDetails.title} poster`);

        expect(img).toBeInTheDocument();
        expect(img).toHaveAttribute("src", mockMovieDetails.imageUrl);

        expect(screen.getByText(mockMovieDetails.title)).toBeInTheDocument();
        expect(screen.getByText(mockMovieDetails.rating.toString())).toBeInTheDocument();
        expect(screen.getByText(mockMovieDetails.genres.join(", "))).toBeInTheDocument();
        expect(screen.getByText(mockMovieDetails.releaseYear.toString())).toBeInTheDocument();
        expect(screen.getByText(mockMovieDetails.description)).toBeInTheDocument();

        const expectedDuration = formatDuration(mockMovieDetails.duration);

        expect(screen.getByText(expectedDuration)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should render different movie details correctly", () => {
        const { asFragment } = renderComponent(alternativeMovieDetails);

        expect(screen.getByAltText(`${alternativeMovieDetails.title} poster`)).toBeInTheDocument();
        expect(screen.getByText(alternativeMovieDetails.title)).toBeInTheDocument();
        expect(screen.getByText(alternativeMovieDetails.description)).toBeInTheDocument();
        expect(screen.getByText(formatDuration(alternativeMovieDetails.duration))).toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot();
    });
});
