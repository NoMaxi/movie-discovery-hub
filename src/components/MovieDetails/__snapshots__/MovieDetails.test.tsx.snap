// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Should render different movie details correctly 1`] = `
<DocumentFragment>
  <div
    class="
                movie-details flex w-full max-h-header-height-expanded gap-x-14 p-8 text-text
                bg-content-background
            "
    data-testid="movie-details"
  >
    <img
      alt="Another Great Film poster"
      class="movie-details-poster flex-shrink-0 w-[320px] h-[485px] object-cover"
      src="https://example.com/poster-alt.jpg"
    />
    <div
      class="movie-details-info flex flex-col flex-grow"
    >
      <div
        class="flex items-center gap-x-5 mb-2"
      >
        <h1
          class="movie-details-title text-2xl font-light tracking-[1px] uppercase"
        >
          Another Great Film
        </h1>
         
        <div
          class="
                            movie-details-rating flex flex-shrink-0 items-center justify-center
                            w-3xl h-3xl border-2 rounded-full
                        "
        >
          <span
            class="font-light"
          >
            7.9
          </span>
        </div>
      </div>
      <p
        class="movie-details-genres text-md opacity-50 mb-6"
      >
        Comedy, Crime
      </p>
      <div
        class="flex items-center gap-x-6 text-[24px] font-light text-primary mb-6"
      >
        <span
          class="movie-details-date"
        >
          2022
        </span>
        <span
          class="movie-details-duration"
        >
          60min
        </span>
      </div>
      <p
        class="
                    movie-details-description font-light opacity-50 leading-relaxed overflow-y-scroll"
      >
        The greatest movie in history of cinema.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MovieDetails Should render movie details correctly with provided data 1`] = `
<DocumentFragment>
  <div
    class="
                movie-details flex w-full max-h-header-height-expanded gap-x-14 p-8 text-text
                bg-content-background
            "
    data-testid="movie-details"
  >
    <img
      alt="Pulp Fiction poster"
      class="movie-details-poster flex-shrink-0 w-[320px] h-[485px] object-cover"
      src="https://upload.wikimedia.org/wikipedia/en/3/3b/Pulp_Fiction_%281994%29_poster.jpg"
    />
    <div
      class="movie-details-info flex flex-col flex-grow"
    >
      <div
        class="flex items-center gap-x-5 mb-2"
      >
        <h1
          class="movie-details-title text-2xl font-light tracking-[1px] uppercase"
        >
          Pulp Fiction
        </h1>
         
        <div
          class="
                            movie-details-rating flex flex-shrink-0 items-center justify-center
                            w-3xl h-3xl border-2 rounded-full
                        "
        >
          <span
            class="font-light"
          >
            8.9
          </span>
        </div>
      </div>
      <p
        class="movie-details-genres text-md opacity-50 mb-6"
      >
        Comedy, Crime
      </p>
      <div
        class="flex items-center gap-x-6 text-[24px] font-light text-primary mb-6"
      >
        <span
          class="movie-details-date"
        >
          1994
        </span>
        <span
          class="movie-details-duration"
        >
          2h 34min
        </span>
      </div>
      <p
        class="
                    movie-details-description font-light opacity-50 leading-relaxed overflow-y-scroll"
      >
        Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra
      </p>
    </div>
  </div>
</DocumentFragment>
`;
